aptos-move/aptos-intent/src/builder.rs

```diff
@@ -0,0 +1,473 @@
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub struct BatchArgumentWASM {
+    ty: BatchArgumentType,
+    raw: Option<Vec<u8>>,
+    signer: Option<u16>,
+    previous_result: Option<PreviousResult>,
+}
+
+#[wasm_bindgen]
+pub struct BatchedFunctionCallBuilder {
+    calls: Vec<BatchedFunctionCall>,
+    num_signers: u16,
+    modules: BTreeMap<ModuleId, CompiledModule>,
+    type_of_returns: BTreeMap<(u16, u16), (TypeTag, AbilitySet)>,
+}
+
+#[wasm_bindgen]
+impl BatchedFunctionCallBuilder {
+    pub fn single_signer() -> Self {
+        Self {
+            calls: vec![],
+            num_signers: 1,
+            modules: BTreeMap::new(),
+            type_of_returns: BTreeMap::new(),
+        }
+    }
+
+    pub fn multi_signer(signer_count: u16) -> Self {
+        Self {
+            calls: vec![],
+            num_signers: signer_count,
+            modules: BTreeMap::new(),
+            type_of_returns: BTreeMap::new(),
+        }
+    }
+
+    pub fn add_batched_call(
+        &mut self,
+        module: String,
+        function: String,
+        ty_args: Vec<String>,
+        args: Vec<BatchArgumentWASM>,
+    ) -> Result<Vec<BatchArgumentWASM>, String> {
+        Ok(self
+            .add_batched_call_impl(
+                module,
+                function,
+                ty_args,
+                args.into_iter().map(BatchArgument::from).collect(),
+            )
+            .map_err(|err| format!("{:?}", err))?
+            .into_iter()
+            .map(|arg| arg.into())
+            .collect())
+    }
+
+    pub fn generate_batched_calls(self) -> Result<Vec<u8>, String> {
+        crate::codegen::generate_script_from_batched_calls(
+            &self.calls,
+            self.num_signers,
+            &self.modules,
+        )
+        .map_err(|err| format!("{:?}", err))
+    }
+}
+
+fn find_function<'a>(
+    map: &'a BTreeMap<ModuleId, CompiledModule>,
+    module_id: &ModuleId,
+    func_name: &IdentStr,
+) -> anyhow::Result<(&'a CompiledModule, &'a FunctionHandle)> {
+    if let Some(module) = map.get(module_id) {
+        for def in module.function_defs() {
+            let handle = module.function_handle_at(def.function);
+            if module.identifier_at(handle.name) == func_name {
+                return Ok((module, handle));
+            }
+        }
+    }
+    anyhow::bail!("{}::{} doesn't exist on chain", module_id, func_name)
+}
+
+impl BatchedFunctionCallBuilder {
+    fn return_values(
+        &mut self,
+        module_id: &ModuleId,
+        func_name: &IdentStr,
+        ty_args: &[TypeTag],
+    ) -> anyhow::Result<Vec<BatchArgument>> {
+        let (module, handle) = find_function(&self.modules, module_id, func_name)?;
+        let mut returns = vec![];
+        for (idx, sig) in module.signature_at(handle.return_).0.iter().enumerate() {
+            let call_idx = (self.calls.len() - 1) as u16;
```

Comment:
if `self.calls.len()` is 0 this underflows and behavior differs between debug and release mode.  Not a great idea.

Reply:
return_values will always be called after `self.calls` is pushed so we should be fine. With that said tho it's always good to add an extra check here.